### DISCUSSION TITLE
New twig filters (deep_merge)

### DISCUSSION THREAD

**Discussion by johanneslamers** [craftcms/cms#8385] [Ideas]
### Description
Came across this link: [https://davidwalsh.name/javascript-deep-merge](https://davidwalsh.name/javascript-deep-merge). 

**Deep merge** can recursively merge any number of objects (including arrays) into one final object. Would be a nice addition to have this as Twig filter. 

Some other twig filters that could be handy:
http://www.aolpublishers.com/support/documentation/sites/filebased/codeblocks/funcs/merge.md
- deep_merge
- merge_replace
- deep_merge_replace

### Additional info
- Craft version: 3

**Comment by brandonkelly** [comment]
Just jotting down some notes in case we/someone else decides to run with this…

- `|deep_merge` – similar behavior to [`array_merge_recursive()`](http://php.net/manual/en/function.array-merge-recursive.php), except that numeric array keys are treated the same as string keys (`array_merge_recursive()` will _append_ numerically-indexed elements to the end of the merged array; `|deep_merge` will merge same-number keys together.)

- `|merge_replace` – similar behavior to Twig’s bulit-in [`|merge`](http://twig.sensiolabs.org/doc/2.x/filters/merge.html) filter (and [`array_merge()`](http://php.net/manual/en/function.array-merge.php)), except that numeric keys are treated the same as string keys.

- `|deep_merge_replace` – similar behavior to Yii’s [`ArrayHelper::merge()`](http://www.yiiframework.com/doc-2.0/yii-helpers-basearrayhelper.html#merge()-detail), except (once again) numeric keys are treated the same as string keys.

**Comment by joshuapease** [comment]
Joining in on this ticket to say we had a similar need in a project and ended up using `ArrayHelper::merge()` to make our own `merge_deep` filter. 

It greatly simplified the default attributes in some form input macros we were building. Having something like this built in to Craft would be super helpful. 

I would be happy to put together a PR if it's something the Craft team is open to. 

#### Exmple: 
```twig

{% macro numberField(props={}) %}
    {% set props = {
        fieldEl: 'input',
        fieldType: 'number',
        fieldAttrs: {
            value: props.value ?? null,
            min: 0,
        },
    } | merge_deep(props) %}

    {# Input code here #}
{% endmacro %}

```

**Comment by brandonkelly** [comment]
Not sure how this didn’t make it into the changelog or documentation, but we added a `recursive` argument to the `|merge` filter in Craft 3.4, which causes the arrays to get merged via `ArrayHepler::merge()`, satisfying the need for a `|deep_merge_replace` filter.
